packages/core/src/loaders/progressiveLoader.ts

```diff
@@ -0,0 +1,382 @@
+import type {
+  IRetrieveConfiguration,
+  IImage,
+  RetrieveStage,
+  EventTypes,
+  ProgressiveListener,
+} from '../types';
+import sequentialRetrieveConfiguration from './sequentialRetrieveConfiguration';
+import interleavedRetrieveConfiguration from './interleavedRetrieveConfiguration';
+import { loadAndCacheImage, loadImage } from './imageLoader';
+import { triggerEvent, ProgressiveIterator, decimate } from '../utilities';
+import imageLoadPoolManager from '../requestPool/imageLoadPoolManager';
+import { ImageStatus, RequestType, Events } from '../enums';
+import cache from '../cache';
+import eventTarget from '../eventTarget';
+
+export { sequentialRetrieveConfiguration, interleavedRetrieveConfiguration };
+
+type StageStatus = {
+  stageId: string;
+  startTime?: number;
+  stageStartTime?: number;
+  numberOfImages: number;
+  numberOfFailures: number;
+  remaining: number;
+};
+
+/**
+ * A progressive loader is given some number of images to load,
+ * and calls a success or failure callback some number of times in some
+ * ordering, possibly calling back multiple times.
+ * This allows the progressive loader to be configured for different setups
+ * and to return render results for various images.
+ *
+ * When used by the a stack viewport, the progressive loader can return multiple
+ * representations to the viewport, replacing earlier/more lossy versions with better ones.
+ *
+ * When used by a streaming loader, the progressive loader can change the ordering
+ * of the rendering to retrieve high priority images first, and the lower priority
+ * images later to provide a complete final rendering.
+ *
+ * Requests are held in a queue, such that subsequent requests for a given
+ * image can be cancelled or ensured to be not initiated until the higher
+ * priority image sets have been completed.
+ *
+ * This loader is also used for the base streamimg image volume, configured with
+ * a minimal interleaved load order, combined with filling nearby volume slices
+ * on load, resulting in much faster initial apparent display.
+ *
+ * The loader will load images from existing cached images, cached volumes, and
+ * from other nearby images or one or more calls to back end services.
+ *
+ * @param imageIds - the set of images to load.  For a volume, these should be
+ *                   ordered from top to bottom.
+ * @param listener - has success and failure callbacks to listen for image deliver events, and may
+ *                   have a getTargetOptions to get information on the retrieve
+ * @param retrieveOptions - is a set of retrieve options to use
+ */
+export async function load(
+  imageIds: string[],
+  listener: ProgressiveListener,
+  retrieveOptions: IRetrieveConfiguration = interleavedRetrieveConfiguration
+): Promise<unknown> {
+  const displayedIterator = new ProgressiveIterator<void | IImage>('displayed');
+  const imageStatus = new Map<string, ImageStatus>();
+  const stageStatus = new Map<string, StageStatus>();
```

Comment:
i prefer imageStatusMap and stageStatusMap to just make it clear and different from its type (ImageStatus, and StageStatus)